core/drivers/amd/mailbox_driver.c

```
@@ -0,0 +1,212 @@
+
+#include <mailbox_private.h>
+
+#include <drivers/amd/mailbox_driver.h>
+
+static ipi_info_t ipi_info;
+
+static TEE_Result mailbox_call(enum mailbox_api ipi_api, uint32_t blocking)
+{
+	struct thread_smc_args args = {
+		.a0 = PLAT_SMC_SIP_IPI | ipi_api,
+		.a1 = reg_pair_to_64(0, ipi_info.local),
+		.a2 = reg_pair_to_64(0, ipi_info.remote),
+		.a3 = reg_pair_to_64(0, blocking),
+	};
+
+	thread_smccc(&args);
+
+	if (args.a0) {
+		EMSG("Mailbox Call returned with error 0x%lx", args.a0);
+		return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result mailbox_open(uint32_t remote_id, size_t payload_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	ipi_info.remote = remote_id;
+
+	paddr_t req_addr = (ipi_info.buf + \
+			    (remote_id * IPI_OFFSET_MULTIPLIER) + \
+			    IPI_REQ_OFFSET);
+
+	paddr_t rsp_addr = (ipi_info.buf + \
+			    (remote_id * IPI_OFFSET_MULTIPLIER) + \
+			    IPI_RESP_OFFSET);
+
+	ipi_info.req = core_mmu_add_mapping(MEM_AREA_RAM_SEC,
+					    req_addr,
+					    payload_size);
+
+	ipi_info.rsp = core_mmu_add_mapping(MEM_AREA_RAM_SEC,
+					    rsp_addr,
+					    payload_size);
+	if (!ipi_info.req || !ipi_info.rsp)
+		panic();
+
+	mutex_lock(&ipi_info.lock);
+
+	res = mailbox_call(IPI_MAILBOX_OPEN, IPI_BLOCK);
+	if (res != TEE_SUCCESS)
+		EMSG("Failed to open Mailbox for remote id %d", remote_id);
+
+	mutex_unlock(&ipi_info.lock);
+
+	return res;
+}
+
+TEE_Result mailbox_release(uint32_t remote_id, size_t payload_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (ipi_info.remote != remote_id) {
+		EMSG("Mailbox not available for remote id %d", remote_id);
+		return res;
+	}
+
+	mutex_lock(&ipi_info.lock);
+
+	res = mailbox_call(IPI_MAILBOX_RELEASE, IPI_BLOCK);
+	if (res != TEE_SUCCESS) {
+		EMSG("Mailbox release failed for remote id %d", remote_id);
+		goto err;
+	}
+
+	res = core_mmu_remove_mapping(MEM_AREA_RAM_SEC,
+				      ipi_info.req,
+				      payload_size);
+	res |= core_mmu_remove_mapping(MEM_AREA_RAM_SEC,
```

This is a micro-optimization, please report the errors separately.